### DISCUSSION TITLE
I want to use an interpretability algorithm for heterogeneous graphs（like from torch_geometric.explain import Explainer, CaptumExplainer）, how should I do it

### DISCUSSION THREAD

**Discussion by GXFay** [pyg-team/pytorch_geometric#6873] [Q&A]
I want to use an interpretability algorithm for heterogeneous graphs（like from torch_geometric.explain import Explainer, CaptumExplainer）, how should I do it

**Comment by rusty1s** [comment]
Can you clarify on what you mean by "interpretability"? Can't you use use `CaptumExplainer` (which supports heterogeneous graphs)?

**Reply by GXFay** [reply]
when I use CaptumExplainer, it told me (RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.) 
<img width="494" alt="image" src="https://user-images.githubusercontent.com/125689220/223647386-7824671d-e390-4130-983f-24fcca9607da.png">
<img width="502" alt="image" src="https://user-images.githubusercontent.com/125689220/223647439-94398d09-afac-47e1-834b-85de85feb7c0.png">

**Reply by rusty1s** [reply]
Mh, I guess this happens if you have node types/edge types that are not reachable within message passing. @wsad1 Is there a way to fix this?

**Reply by GXFay** [reply]
Thanks for u answer. I try to use to_captum_model, but it still report an error: 
<img width="430" alt="image" src="https://user-images.githubusercontent.com/125689220/223721447-92ca0538-e3a9-4795-a9f1-a6ca561fa384.png">
<img width="553" alt="image" src="https://user-images.githubusercontent.com/125689220/223721519-09993dc1-8de3-47a1-8c1a-47f0aed90630.png">

**Reply by rusty1s** [reply]
How do `node_types` and `x_dict.keys()` look like?

**Comment by yfeng77** [comment]
Hi, I got the same error. Did you find a solution? And when I run 
```
hetero_explanation = explainer(
    data.x_dict,
    data.edge_index_dict,
    index=torch.tensor([1, 3]),
) 
```
again without changing anything, then I got 

> 'IntegratedGradients' object is not callable.

error.

**Reply by rusty1s** [reply]
Do you have a reproducible example? We added an example of using `CaptumExplainer` on heterogeneous graphs in https://github.com/pyg-team/pytorch_geometric/pull/7096.

**Reply by yfeng77** [reply]
> Do you have a reproducible example? We added an example of using `CaptumExplainer` on heterogeneous graphs in #7096.

Hi. Thank you for your response. I am working on the Explaining node classification on a heterogeneous graph that I found on the [PyG tutorial ](https://pytorch-geometric.readthedocs.io/en/latest/tutorial/explain.html). For the hetero_data, I tried the [IMDB dataset](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/han_imdb.py) and [OGB_MAG dataset](https://pytorch-geometric.readthedocs.io/en/latest/tutorial/heterogeneous.html). For model, I tried HeteroGNN, HGT and HAN. But none of them work. When I used the IMDB dataset and the HAN model, I got the same error as the author got. Thank you!

**Reply by rusty1s** [reply]
Do you have a small example to share so that I am able to reproduce?

**Reply by yfeng77** [reply]
> Do you have a small example to share so that I am able to reproduce?

Yes. Hope this example will help. Thank you so much.
[Untitled (1).pdf](https://github.com/pyg-team/pytorch_geometric/files/11211706/Untitled.1.pdf)

**Reply by rusty1s** [reply]
Sorry for super late reply. I assume this is because our heterogeneous explainer is designed with `to_hetero` or `HeteroConv` in mind, and so it fails to explain heterogeneous graphs where there does not exist separate GNN layers for each edge type.
This is temporarily fixable by:
```python
explainer = Explainer(
    model=model,  # It is assumed that model outputs a single tensor.
    algorithm=CaptumExplainer('IntegratedGradients'),
    explanation_type='model',
    node_mask_type='attributes',
    edge_mask_type=None,
    model_config=dict(
        mode='multiclass_classification',
        task_level="node",
        return_type='probs',  # Model returns probabilities.
    ),
)
```
but this is obviously not a great workaround.

**Reply by yfeng77** [reply]
Thank you! I may try another GNN model.
